lib/Conversion/ArcToLLVM/LowerArcToLLVM.cpp

```diff
@@ -285,6 +304,215 @@ struct ReplaceOpWithInputPattern : public OpConversionPattern<OpTy> {
+        modelInfo(modelInfo) {}
+
+protected:
+  Value createPtrToPortState(ConversionPatternRewriter &rewriter, Location loc,
+                             Value state, const StateInfo &port) const {
+    MLIRContext *ctx = rewriter.getContext();
+    return rewriter.create<LLVM::GEPOp>(loc, LLVM::LLVMPointerType::get(ctx),
+                                        IntegerType::get(ctx, 8), state,
+                                        LLVM::GEPArg(port.offset));
+  }
+
+  llvm::DenseMap<StringRef, ModelInfoMap> &modelInfo;
+};
+
+/// Lowers SimInstantiateOp to a malloc and memset call. This pattern will
+/// mutate the global module.
+struct SimInstantiateOpLowering
+    : public ModelAwarePattern<arc::SimInstantiateOp> {
+  using ModelAwarePattern::ModelAwarePattern;
+
+  LogicalResult
+  matchAndRewrite(arc::SimInstantiateOp op, OpAdaptor adaptor,
+                  ConversionPatternRewriter &rewriter) const final {
+    auto modelIt = modelInfo.find(
+        cast<SimModelInstanceType>(op.getBody().getArgument(0).getType())
+            .getModel());
+    ModelInfoMap &model = modelIt->second;
+
+    ModuleOp moduleOp = op->getParentOfType<ModuleOp>();
+    if (!moduleOp)
+      return failure();
+
+    ConversionPatternRewriter::InsertionGuard guard(rewriter);
+
+    Type convertedIndex = typeConverter->convertType(rewriter.getIndexType());
+
+    LLVM::LLVMFuncOp mallocFunc =
+        LLVM::lookupOrCreateMallocFn(moduleOp, convertedIndex);
+    LLVM::LLVMFuncOp freeFunc = LLVM::lookupOrCreateFreeFn(moduleOp);
+
+    Location loc = op.getLoc();
+    Value numStateBytes = rewriter.create<LLVM::ConstantOp>(
+        loc, convertedIndex, model.numStateBytes);
+    Value allocated =
+        rewriter
+            .create<LLVM::CallOp>(loc, mallocFunc, ValueRange{numStateBytes})
+            .getResult();
+    Value zero =
+        rewriter.create<LLVM::ConstantOp>(loc, rewriter.getI8Type(), 0);
+    rewriter.create<LLVM::MemsetOp>(loc, allocated, zero, numStateBytes, false);
+    rewriter.inlineBlockBefore(&adaptor.getBody().getBlocks().front(), op,
+                               {allocated});
+    rewriter.create<LLVM::CallOp>(loc, freeFunc, ValueRange{allocated});
+    rewriter.eraseOp(op);
+
+    return success();
+  }
+};
+
+struct SimSetInputOpLowering : public ModelAwarePattern<arc::SimSetInputOp> {
+  using ModelAwarePattern::ModelAwarePattern;
+
+  LogicalResult
+  matchAndRewrite(arc::SimSetInputOp op, OpAdaptor adaptor,
+                  ConversionPatternRewriter &rewriter) const final {
+    auto modelIt = modelInfo.find(
+        cast<SimModelInstanceType>(op.getInstance().getType()).getModel());
+    ModelInfoMap &model = modelIt->second;
+
+    auto portIt = model.states.find(op.getInput());
+    if (portIt == model.states.end()) {
+      // If the port is not found in the state, it means the model does not
+      // actually use it. Thus this operation is a no-op.
+      rewriter.eraseOp(op);
+      return success();
+    }
```

Comment:
This means that the `arc.root_input` operation was optimized away, but I'm wondering whether we should even allow this. I'd say the model should always reserve a slot for every input but, of course, it is free to just not use it.
This is not an issue of this PR though. Handling it like this is great for now.